packages/hadron-build/lib/signtool.js

```diff
@@ -1,22 +1,103 @@
-const debug = require('debug')('hadron-build:signtool');
-const { execFileSync } = require('child_process');
 const path = require('path');
+const debug = require('debug')('hadron-build:target');
+const { sign: _garasign } = require('@mongodb-js/signing-utils');
 
-async function signtool(fileToSign) {
-  const signtoolPath = path.resolve(__dirname, '..', 'signtool/signtool.exe');
-
-  const execArgs = [signtoolPath, [path.resolve(fileToSign)], { stdio: 'inherit' }];
-  debug(`Running signtool.exe to sign '${signtoolPath}'`, {
-    execArgs: execArgs,
-    env: {
-      NOTARY_SIGNING_COMMENT: process.env.NOTARY_SIGNING_COMMENT,
-      NOTARY_URL: process.env.NOTARY_URL,
-      NOTARY_SIGNING_KEY: process.env.NOTARY_SIGNING_KEY,
-    }
-  });
+const canSign = () => (
+  process.env.GARASIGN_USERNAME &&
+  process.env.GARASIGN_PASSWORD &&
+  process.env.ARTIFACTORY_USERNAME &&
+  process.env.ARTIFACTORY_PASSWORD
+);
+
+/**
+ * When using gpg to sign a file, it creates a signature file
+ * with same name as the original file and adds `.sig` to it.
+ *
+ * @param {string} filename
+ * @returns string
+ */
+function getSignedFilename(filename) {
+  return `${filename}.sig`;
+}
+
+/**
+ * Currently, windows and macos zip are created from `zip` step
+ * of the release process and we sign them here. For linux, we
+ * create and sign the archive when creating corresponding deb/rpm file.
+ *
+ * @param {import('./Target')} target
+ */
+function signArchive(target, cb) {
+  const { app_archive_name, platform } = target;
+  if (platform === 'linux') {
+    debug('linux archive is signed when creating deb/rpm');
+    return cb();
+  }
+  sign(target.dest(app_archive_name)).then(cb).catch(cb);
+}
+
+/**
+ *
+ * For signing a file, we have following options with the current CI setup:
+ *
+ * |----------|--------|--------|----------------|--------------|
+ * | platform | docker | client | signing method | filetype     |
+ * |----------|--------|--------|----------------|--------------|
+ * | macOS    | no     | remote | gpg            | zip          |
+ * | ubuntu   | yes    | local  | gpg            | zip, deb     |
+ * | rhel     | yes ‡  | remote | gpg            | zip, rpm     |
+ * | windows  | no     | remote | jsign          | msi, exe     |
+ * | windows  | no     | remote | gpg            | zip, nupkg   |
+ * |----------|--------|--------|----------------|--------------|
+ *
+ * ‡ On rhel we have docker, but currently we are using rhel7.6
+ *  for packaging and docker does not support login using --password-stdin
+ *
+ * @param {string} src
+ * @returns {Promise<void>}
+ */
+async function sign(src, garasign = _garasign) {
+  const variant = process.env.EVERGREEN_BUILD_VARIANT;
```

Comment:
We have similar branching based on build variants in mongosh, and personally i find hard to digest and maintain, it creates obscure dependencies between the CI and the code that may not be immediately clear, and relies on assumptions about the variants that may not stand true over time.

Could we do something different to detect where to use ssh and where to use docker? For example, we could directly set this as a variable / expansion in CI: https://docs.devprod.prod.corp.mongodb.com/evergreen/Project-Configuration/Project-Configuration-Files#build-variants

```
name: ubuntu
expansions:
    garasign_client_type: "remote"
```